File: internal/desktop/frontend/eslint-rules/policy.js
import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";

const policyPath = path.resolve(path.dirname(fileURLToPath(import.meta.url)), "../../../../comment-discipline.json");

export function loadPolicy(file = policyPath) {
  let value;
  try {
    value = JSON.parse(fs.readFileSync(file, "utf8"));
  } catch (error) {
    throw new Error(`read policy: ${error.message}`, { cause: error });
  }

  if (!value || typeof value !== "object" || Array.isArray(value)) {
    throw new Error("policy must be a JSON object");
  }
  const keys = Object.keys(value).sort();
  const expected = ["maxCommentRun", "narrationPhrases", "pathExtensions", "schemaVersion"];
  if (keys.length !== expected.length || keys.some((key, index) => key !== expected[index])) {
    throw new Error(`policy must contain exactly: ${expected.join(", ")}`);
  }
  if (value.schemaVersion !== 1) throw new Error("policy schemaVersion must be 1");
  if (!Number.isInteger(value.maxCommentRun) || value.maxCommentRun < 1) {
    throw new Error("policy maxCommentRun must be a positive integer");
  }
  validateList(value.narrationPhrases, "narrationPhrases");
  validateList(value.pathExtensions, "pathExtensions", true);
  return value;
}

function validateList(values, name, extension = false) {
  if (!Array.isArray(values) || values.length === 0) throw new Error(`policy ${name} must not be empty`);
  const seen = new Set();
  for (const value of values) {
    if (typeof value !== "string" || value === "" || value.trim() !== value || value !== value.toLowerCase()) {
      throw new Error(`policy ${name} entries must be non-empty, trimmed, and lowercase`);
    }
    if (extension && (value.startsWith(".") || value.includes("/") || value.includes("\\"))) {
      throw new Error("policy pathExtensions entries must omit dots and path separators");
    }
    if (seen.has(value)) throw new Error(`policy ${name} contains duplicate ${JSON.stringify(value)}`);
    seen.add(value);
  }
}

File: internal/commentdiscipline/policy.go
package commentdiscipline

import (
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"os"
	"strings"
)

type Policy struct {
	SchemaVersion    int      `json:"schemaVersion"`
	MaxCommentRun    int      `json:"maxCommentRun"`
	NarrationPhrases []string `json:"narrationPhrases"`
	PathExtensions   []string `json:"pathExtensions"`
}

func LoadPolicy(path string) (Policy, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return Policy{}, fmt.Errorf("read policy: %w", err)
	}
	decoder := json.NewDecoder(strings.NewReader(string(data)))
	decoder.DisallowUnknownFields()
	var policy Policy
	if err := decoder.Decode(&policy); err != nil {
		return Policy{}, fmt.Errorf("decode policy: %w", err)
	}
	var trailing any
	if err := decoder.Decode(&trailing); !errors.Is(err, io.EOF) {
		if err == nil {
			return Policy{}, fmt.Errorf("decode policy: multiple JSON values")
		}
		return Policy{}, fmt.Errorf("decode policy: %w", err)
	}
	if policy.SchemaVersion != 1 {
		return Policy{}, fmt.Errorf("schemaVersion must be 1")
	}
	if policy.MaxCommentRun < 1 {
		return Policy{}, fmt.Errorf("maxCommentRun must be at least 1")
	}
	if err := validateList("narrationPhrases", policy.NarrationPhrases, false); err != nil {
		return Policy{}, err
	}
	if err := validateList("pathExtensions", policy.PathExtensions, true); err != nil {
		return Policy{}, err
	}
	return policy, nil
}

func validateList(name string, values []string, extension bool) error {
	if len(values) == 0 {
		return fmt.Errorf("%s must not be empty", name)
	}
	seen := make(map[string]struct{}, len(values))
	for _, value := range values {
		if value == "" || value != strings.TrimSpace(value) || value != strings.ToLower(value) {
			return fmt.Errorf("%s entries must be non-empty, trimmed, and lowercase", name)
		}
		if extension && (strings.HasPrefix(value, ".") || strings.ContainsAny(value, `/\\`)) {
			return fmt.Errorf("pathExtensions entries must omit dots and path separators")
		}
		if _, ok := seen[value]; ok {
			return fmt.Errorf("%s contains duplicate %q", name, value)
		}
		seen[value] = struct{}{}
	}
	return nil
}
